src/VASSAL/counters/BasicPiece.java

```diff
@@ -571,6 +612,13 @@ public String getState() {
     final Point p = getPosition();
     se.append(p.x).append(p.y);
     se.append(getGpId());
+    se.append(persistentProps.size());
+    // Persistent Property values will always be String (for now).
+    for (Object key : persistentProps.keySet()) {
+      se.append(key.toString()); 
+      String val = (String) persistentProps.get(key);
+      se.append(key == null ? "" : val);
```

Comment:
key can never be null here else it would have thrown an NPE above

Reply:
Thanks, same bug. Fixed whole section as per your suggestion
    persistentProps.forEach((key, val) -> {
      se.append(key == null ? "" : key.toString());
      se.append(val == null ? "" : val.toString());
    });
Much neater.